Miscellany

b20n opened the pull request.

b20n committed:
Use gen_fsm:reply/2, not gen_server:reply/2, from kofta_metadata

b20n committed:
Remove kofta_producer_batcher_sup - it's unused

b20n committed:
Use a nonblocking poolboy pattern in kofta_connection

b20n committed:
Cleanly crash producer_batcher on connection failure

We don't want to attempt a retry here, because timeout logic gets complicated
and it's better to ensure exactly-once delivery.

b20n committed:
Be more explicit about request timeouts in producer_batcher

This is an effort to ensure sane semantics in the case of failures

b20n committed:
Handle connection request errors in kofta:fetch/3

sagelywizard commented on src/kofta_metadata.erl:
```diff
@@ -165,7 +167,7 @@ disconnected(_Msg, _From, State) ->
 
-    NewState = State#st{last_reconnect=os:timestamp()},
+    NewState = State#st{last_reconnect=os:timestamp(), clients=Clients1},
     {reply, {error, broker_down}, NextState, NewState, Timeout}.
```
You'll want to not reply here if you add it to the clients dict.

sagelywizard replied:
Also, not returning an error will cause memory usage to grow unboundedly for each call to metadata when the client is disconnected. I had it return an error rather than add it to the client dict because there could potentially be a lot of lookup calls. But FWIW I didn't put much thought into if it actually could be a problem.

b20n replied:
Bah, I misread the return and thought it was silently dropping the message. I'm +1 on the `{error, broker_down}` reply.

sagelywizard replied:
Sounds good to me.

b20n committed:
Handle lists:keyfind error when bad partition is specified

sagelywizard commented:
NVM

sagelywizard commented on src/kofta_decode.erl:
```diff
@@ -154,30 +163,26 @@ message_set_int(Binary, RestCount, Acc0) ->
+                        Key,
+                        Value
+                    },
+                    message_set_int(Rest2, [Message|Acc0])
```
Err, there is no `message_set_int/2`.

b20n replied:
See previous comment about broken revision...

sagelywizard commented on src/kofta_decode.erl:
```diff
@@ -146,7 +155,7 @@ partition_metadata(Binary) ->
     message_set_int(Binary, Size, []).
 
-message_set_int(Binary, 0, Acc) ->
+message_set_int(Binary, _RestCount, Acc) when bit_size(Binary) < 144 ->
```
144?

b20n replied:
It's the sum of the bits in the first match.

sagelywizard commented on src/kofta_decode.erl:
```diff
@@ -154,30 +163,26 @@ message_set_int(Binary, RestCount, Acc0) ->
-      KeyLen:32/big-signed-integer,
       Rest0/binary>> = Binary,
-
-    {Key, SizeCount0, ValueLen, Rest2} = case KeyLen of
```
The reason why I didn't use `bytes/1` here is because `message_set_int/3` needs to count how many bytes have been decoded.

b20n replied:
Yeah, sorry - I pushed a broken revision of this. I'll push a fixed up one soonish (not tonight).

b20n replied:
Alright, this is fixed up. `message_set_int` doesn't need to keep track of how many bytes have been decoded. That's an incidental change, though - the real goal is/was to avoid causing the whole thing to blow up when passed truncated input.

I think it's probably worth making another pass at the whole thing and ensuring that we're not going to crash with `badmatch` on truncated input. This was an easy place to trigger that sort of failure in the existing code, but I suspect there are more. We'll also want to pass `undefined` values back up the stack. For example, `kofta_fetch:decode/1` needs to handle `undefined` results from `kofta_decode:message_set/1`.

b20n committed:
Rework message_set decoding to prevent badmatch on truncation

The previous implementation required incoming message sets to be
well-formed. Any message set truncated halfway through would cause
the various binary parses to badmatch.
